polar2grid/writers/hdf5.py

```diff
@@ -0,0 +1,302 @@
+from __future__ import annotations
+from typing import TextIO
+
+import os
+import sys
+import logging
+
+import pkg_resources
+import satpy
+import h5py
+import numpy as np
+import xarray as xr
+import dask.array as da
+
+from itertools import groupby
+from datetime import datetime as datetime
+from satpy.dataset import DataID
+from satpy.writers import ImageWriter
+from satpy.writers import compute_writer_results
+
+from trollsift import Parser
+
+from pyresample.geometry import SwathDefinition
+from polar2grid.utils.legacy_compat import convert_p2g_pattern_to_satpy
+from polar2grid.writers.geotiff import NUMPY_DTYPE_STRS, NumpyDtypeList, str_to_dtype, int_or_float
+
+from polar2grid.core.script_utils import setup_logging, rename_log_file, create_exc_handler
+
+USE_POLAR2GRID_DEFAULTS = bool(int(os.environ.setdefault("USE_POLAR2GRID_DEFAULTS", "1")))
+
+LOG = logging.getLogger(__name__)
+
+# reader_name -> filename
+DEFAULT_OUTPUT_FILENAMES = {None: "{platform_name}_{sensor}_{start_time:%Y%m%d_%H%M%S}.h5"}
+
+
+def dist_is_editable(dist) -> bool:
+    """Is distribution an editable install?"""
+    for path_item in sys.path:
+        egg_link = os.path.join(path_item, dist.project_name + ".egg-link")
+        if os.path.isfile(egg_link):
+            return True
+    return False
+
+
+dist = pkg_resources.get_distribution("polar2grid")
+if dist_is_editable(dist):
+    p2g_etc = os.path.join(dist.module_path, "etc")
+else:
+    p2g_etc = os.path.join(sys.prefix, "etc", "polar2grid")
+
+config_path = satpy.config.get("config_path")
+if p2g_etc not in config_path:
+    satpy.config.set(config_path=config_path + [p2g_etc])
+
+
+def all_equal(iterable: list[str]) -> bool:
+    "Returns True if all the elements are equal to each other"
+    g = groupby(iterable)
+    return next(g, True) and not next(g, False)
+
+
+class FakeHDF5:
+    def __init__(self, fh : TextIO[IO[AnyStr]], output_filename: str, var_name: str, compression: bool):
```

Comment:
One of the important parts of this workaround (passing this FakeHDF5 instance) is that we don't pass anything that resembles an open file to dask. If I'm reading this correctly `fh` is the actual h5py File object, right? Is that needed? Can the dataset be created before hand?

Reply:
Okay.  I am a little confused on this one.  Yes, I can pass the dask array and the closed file handle to dask?  Or, are you saying I should pass the dask array and the hdf5 variable name to dask?  Though we are trying to avoid passing an open file handle, we are trying to pass a dask array that is either delayed or computed, correct?

Reply:
The `FakeHDF5` file `__init__` method should only get non-dask and non-open-file objects so strings (filenames, variable names, etc) and other configuration options like booleans. We want to avoid `FakeHDF5` holding on to open file objects (like the h5py File object or other h5py objects) as they can't be serialized which dask will need to do when passing it around in certain situations. Computed numpy arrays would be fine, but I'm not sure you need them in this writer.

The `__setitem__` is a different story as dask will already have serialized and computed things for you.

Reply:
I think I have this resolved in b56af2cc.  The FakeHDF5 class constructor only accepts a filename and var_name.  The ```__setitem__``` writes by opening a file handle to the filename which already has the datasets defined, just not filled with data.  b56af2cc